File: backend/app/models/product.py
"""Product, Category, Variant, Image models."""
import uuid
from datetime import datetime
from sqlalchemy import (
    String, Boolean, DateTime, Float, Text, Integer,
    ForeignKey, Index, Enum as SAEnum,
)
from sqlalchemy.dialects.postgresql import UUID, JSONB, TSVECTOR
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base
import enum


class ProductStatus(str, enum.Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    OUT_OF_STOCK = "out_of_stock"
    DISCONTINUED = "discontinued"


class UnitType(str, enum.Enum):
    KG = "kg"
    GRAM = "gram"
    LITRE = "litre"
    ML = "ml"
    PIECE = "piece"
    DOZEN = "dozen"
    PACK = "pack"


class ProductCategory(Base):
    __tablename__ = "product_categories"
    __table_args__ = (
        Index("ix_product_categories_parent_id", "parent_id"),
        Index("ix_product_categories_slug", "slug", unique=True),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    slug: Mapped[str] = mapped_column(String(120), unique=True, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    icon_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    image_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    parent_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("product_categories.id"), nullable=True
    )
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )

    # Self-referential relationship
    children: Mapped[list["ProductCategory"]] = relationship(
        back_populates="parent", lazy="selectin"
    )
    parent: Mapped["ProductCategory | None"] = relationship(
        back_populates="children", remote_side="ProductCategory.id", lazy="selectin"
    )
    products: Mapped[list["Product"]] = relationship(
        back_populates="category", lazy="selectin"
    )


class Product(Base):
    __tablename__ = "products"
    __table_args__ = (
        Index("ix_products_vendor_id", "vendor_id"),
        Index("ix_products_category_id", "category_id"),
        Index("ix_products_status", "status"),
        Index("ix_products_slug", "slug"),
        Index("ix_products_price", "price"),
        Index("ix_products_search", "search_vector", postgresql_using="gin"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    vendor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vendors.id", ondelete="CASCADE"), nullable=False
    )
    category_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("product_categories.id"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    slug: Mapped[str] = mapped_column(String(250), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    short_description: Mapped[str | None] = mapped_column(String(500), nullable=True)

    # Pricing
    price: Mapped[float] = mapped_column(Float, nullable=False)
    compare_at_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    cost_price: Mapped[float | None] = mapped_column(Float, nullable=True)

    # Inventory
    sku: Mapped[str | None] = mapped_column(String(50), nullable=True)
    barcode: Mapped[str | None] = mapped_column(String(50), nullable=True)
    stock_quantity: Mapped[int] = mapped_column(Integer, default=0)
    low_stock_threshold: Mapped[int] = mapped_column(Integer, default=5)
    track_inventory: Mapped[bool] = mapped_column(Boolean, default=True)

    # Unit
    unit_type: Mapped[UnitType] = mapped_column(
        SAEnum(UnitType), default=UnitType.KG, nullable=False
    )
    unit_value: Mapped[float] = mapped_column(Float, default=1.0)

    # Status & Flags
    status: Mapped[ProductStatus] = mapped_column(
        SAEnum(ProductStatus), default=ProductStatus.ACTIVE, nullable=False
    )
    is_featured: Mapped[bool] = mapped_column(Boolean, default=False)
    is_organic: Mapped[bool] = mapped_column(Boolean, default=False)

    # Ratings
    avg_rating: Mapped[float] = mapped_column(Float, default=0.0)
    total_reviews: Mapped[int] = mapped_column(Integer, default=0)
    total_sold: Mapped[int] = mapped_column(Integer, default=0)

    # Search
    search_vector: Mapped[str | None] = mapped_column(TSVECTOR, nullable=True)
    tags: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    nutritional_info: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    # Timestamps
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow
    )

    # Relationships
    vendor: Mapped["Vendor"] = relationship(back_populates="products")  # noqa: F821
    category: Mapped["ProductCategory"] = relationship(back_populates="products")
    variants: Mapped[list["ProductVariant"]] = relationship(
        back_populates="product", cascade="all, delete-orphan", lazy="selectin"
    )
    images: Mapped[list["ProductImage"]] = relationship(
        back_populates="product", cascade="all, delete-orphan", lazy="selectin"
    )
    reviews: Mapped[list["Review"]] = relationship(  # noqa: F821
        back_populates="product", lazy="selectin"
    )


class ProductVariant(Base):
    __tablename__ = "product_variants"
    __table_args__ = (
        Index("ix_product_variants_product_id", "product_id"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id", ondelete="CASCADE"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    sku: Mapped[str | None] = mapped_column(String(50), nullable=True)
    price: Mapped[float] = mapped_column(Float, nullable=False)
    compare_at_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    stock_quantity: Mapped[int] = mapped_column(Integer, default=0)
    unit_type: Mapped[UnitType] = mapped_column(
        SAEnum(UnitType), default=UnitType.KG, nullable=False
    )
    unit_value: Mapped[float] = mapped_column(Float, default=1.0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    attributes: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )

    product: Mapped["Product"] = relationship(back_populates="variants")


class ProductImage(Base):
    __tablename__ = "product_images"
    __table_args__ = (
        Index("ix_product_images_product_id", "product_id"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id", ondelete="CASCADE"), nullable=False
    )
    image_url: Mapped[str] = mapped_column(String(500), nullable=False)
    alt_text: Mapped[str | None] = mapped_column(String(200), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    is_primary: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )

    product: Mapped["Product"] = relationship(back_populates="images")

File: frontend/lib/models/product.dart
class ProductCategory {
  final String id;
  final String name;
  final String slug;
  final String? description;
  final String? iconUrl;
  final String? imageUrl;
  final String? parentId;
  final int sortOrder;
  final bool isActive;

  ProductCategory({
    required this.id,
    required this.name,
    required this.slug,
    this.description,
    this.iconUrl,
    this.imageUrl,
    this.parentId,
    this.sortOrder = 0,
    this.isActive = true,
  });

  factory ProductCategory.fromJson(Map<String, dynamic> json) {
    return ProductCategory(
      id: json['id'],
      name: json['name'],
      slug: json['slug'],
      description: json['description'],
      iconUrl: json['icon_url'],
      imageUrl: json['image_url'],
      parentId: json['parent_id'],
      sortOrder: json['sort_order'] ?? 0,
      isActive: json['is_active'] ?? true,
    );
  }
}

class ProductImage {
  final String id;
  final String imageUrl;
  final String? altText;
  final int sortOrder;
  final bool isPrimary;

  ProductImage({
    required this.id,
    required this.imageUrl,
    this.altText,
    this.sortOrder = 0,
    this.isPrimary = false,
  });

  factory ProductImage.fromJson(Map<String, dynamic> json) {
    return ProductImage(
      id: json['id'],
      imageUrl: json['image_url'],
      altText: json['alt_text'],
      sortOrder: json['sort_order'] ?? 0,
      isPrimary: json['is_primary'] ?? false,
    );
  }
}

class ProductVariant {
  final String id;
  final String name;
  final String? sku;
  final double price;
  final double? compareAtPrice;
  final int stockQuantity;
  final String unitType;
  final double unitValue;
  final bool isActive;
  final Map<String, dynamic>? attributes;

  ProductVariant({
    required this.id,
    required this.name,
    this.sku,
    required this.price,
    this.compareAtPrice,
    this.stockQuantity = 0,
    required this.unitType,
    this.unitValue = 1.0,
    this.isActive = true,
    this.attributes,
  });

  factory ProductVariant.fromJson(Map<String, dynamic> json) {
    return ProductVariant(
      id: json['id'],
      name: json['name'],
      sku: json['sku'],
      price: (json['price'] as num).toDouble(),
      compareAtPrice: json['compare_at_price']?.toDouble(),
      stockQuantity: json['stock_quantity'] ?? 0,
      unitType: json['unit_type'] ?? 'kg',
      unitValue: (json['unit_value'] as num?)?.toDouble() ?? 1.0,
      isActive: json['is_active'] ?? true,
      attributes: json['attributes'],
    );
  }
}

class Product {
  final String id;
  final String vendorId;
  final String categoryId;
  final String name;
  final String slug;
  final String? description;
  final String? shortDescription;
  final double price;
  final double? compareAtPrice;
  final String? sku;
  final int stockQuantity;
  final String unitType;
  final double unitValue;
  final String status;
  final bool isFeatured;
  final bool isOrganic;
  final double avgRating;
  final int totalReviews;
  final int totalSold;
  final Map<String, dynamic>? tags;
  final Map<String, dynamic>? nutritionalInfo;
  final List<ProductImage> images;
  final List<ProductVariant> variants;
  final DateTime createdAt;
  final DateTime updatedAt;

  Product({
    required this.id,
    required this.vendorId,
    required this.categoryId,
    required this.name,
    required this.slug,
    this.description,
    this.shortDescription,
    required this.price,
    this.compareAtPrice,
    this.sku,
    this.stockQuantity = 0,
    required this.unitType,
    this.unitValue = 1.0,
    required this.status,
    this.isFeatured = false,
    this.isOrganic = false,
    this.avgRating = 0.0,
    this.totalReviews = 0,
    this.totalSold = 0,
    this.tags,
    this.nutritionalInfo,
    this.images = const [],
    this.variants = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  String get primaryImageUrl {
    final primary = images.where((img) => img.isPrimary).firstOrNull;
    return primary?.imageUrl ?? (images.isNotEmpty ? images.first.imageUrl : '');
  }

  bool get isOnSale =>
      compareAtPrice != null && compareAtPrice! > price;

  double get discountPercentage {
    if (!isOnSale) return 0;
    return ((compareAtPrice! - price) / compareAtPrice! * 100);
  }

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'],
      vendorId: json['vendor_id'],
      categoryId: json['category_id'],
      name: json['name'],
      slug: json['slug'],
      description: json['description'],
      shortDescription: json['short_description'],
      price: (json['price'] as num).toDouble(),
      compareAtPrice: json['compare_at_price']?.toDouble(),
      sku: json['sku'],
      stockQuantity: json['stock_quantity'] ?? 0,
      unitType: json['unit_type'] ?? 'kg',
      unitValue: (json['unit_value'] as num?)?.toDouble() ?? 1.0,
      status: json['status'] ?? 'active',
      isFeatured: json['is_featured'] ?? false,
      isOrganic: json['is_organic'] ?? false,
      avgRating: (json['avg_rating'] as num?)?.toDouble() ?? 0.0,
      totalReviews: json['total_reviews'] ?? 0,
      totalSold: json['total_sold'] ?? 0,
      tags: json['tags'],
      nutritionalInfo: json['nutritional_info'],
      images: (json['images'] as List<dynamic>?)
              ?.map((e) => ProductImage.fromJson(e))
              .toList() ??
          [],
      variants: (json['variants'] as List<dynamic>?)
              ?.map((e) => ProductVariant.fromJson(e))
              .toList() ??
          [],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }
}
